### PR TITLE
Add tests for configuration.py

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**
This PR:
- extracts out the casting of strings to known datatypes into a method
- adds several tests for the public methods in `configuration.py`.

A note on the two disabled tests:
Two tests `test_parse_config1` and `test_parse_config2` are disabled.
These tests fail in the github runner.
The code in `configuration.py`, calls `_get_shell_env` which executes in a subprocess.  For some reason, it is executed in `/bin/sh` instead of `/bin/bash`.  `source` is only available in `bash`, and not in `sh`.  If `source` is replaced w/ `.`, this code will succeed in the runner.
On the localhost, these tests run as expected even with `source`, so I think this is something peculiar w/ the runner.
`bash` is available, but `subprocess` is opening a `sh` subprocess.

This was identified in the work when passing `os.environ` to the task e.g. PR #1106 e.g. line https://github.com/NOAA-EMC/global-workflow/blob/bae934f307d9565271f875c31484bc9e759e4666/scripts/exgdas_global_aero_analysis_initialize.py#L12 Currently, all values of the keys passed by `os.environ` will be strings.  Casting them as known datatypes can significantly improve the user experience.

Note:
2 tests for `parse_config` are failing in GH runner, but they pass on localhost.  Currently just marking them to skip till the issue with the runner is identified.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
- [x] Generated experiment directories and XML for various combinations of `cycled`+`ATM` and `forecast-only`+`ATM, S2SWA`.  The resulting experiment directories and XML's were identical.
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [X] Any dependent changes have been merged and published

### CODE REVIEW

File: ush/python/pygw/src/pygw/configuration.py [hunk 1]
```diff
@@ -137,16 +121,59 @@ def _get_shell_env(scripts: List) -> Dict[str, Any]:
...
         except ValueError:
-            return value
+            return string
 
-    @staticmethod
-    def _true_or_not(value):
+    def _true_or_not(string: str):
         try:
-            return value.lower() in Configuration.TRUTHS
+            return string.lower() in TRUTHS
```
Comment (WalterKolczynski-NOAA) [inline ush/python/pygw/src/pygw/configuration.py:167]: We're kinda treating the lists inconsistently. `TRUTHS` gets a single case format and all inputs are converted to lower for comparison, but `BOOLS` gets a number of allowed capitalizations (upper, lower, title) and has to match one of them. Personally prefer the former, so typos like `TRue` would match, but fine with either choice.

File: ush/python/pygw/src/pygw/timetools.py [hunk 2]
```diff
@@ -9,12 +9,15 @@
...
 _DATETIME_RE = re.compile(
-    r"(?P<year>\d{4})(-)?(?P<month>\d{2})(-)?(?P<day>\d{2})(T)?(?P<hour>\d{2})?(:)?(?P<minute>\d{2})?(:)?(?P<second>\d{2})?(Z)?")
+    r"(?P<year>\d{4})(-)?(?P<month>\d{2})(-)?(?P<day>\d{2})"
+    r"(T)?(?P<hour>\d{2})?(:)?(?P<minute>\d{2})?(:)?(?P<second>\d{2})?(Z)?")
 
 _TIMEDELTA_HOURS_RE = re.compile(
-    r"(?P<sign>[+-])?((?P<days>\d+)[d])?(T)?((?P<hours>\d+)[H])?((?P<minutes>\d+)[M])?((?P<seconds>\d+)[S])?(Z)?")
+    r"(?P<sign>[+-])?"
+    r"((?P<days>\d+)[d])?(T)?((?P<hours>\d+)[H])?((?P<minutes>\d+)[M])?((?P<seconds>\d+)[S])?(Z)?")
 _TIMEDELTA_TIME_RE = re.compile(
```
Comment (WalterKolczynski-NOAA) [inline ush/python/pygw/src/pygw/timetools.py:18]: These aren't new, but while I'm noticing it: these two constants are named `HOURS` and `TIME`, but they both take in hours, minutes, and seconds (in different formats). Are there better names to use (or maybe a list of regex under one variable)?

**Review by WalterKolczynski-NOAA — COMMENTED** [review]
I know this is still draft, but a couple notes.

File: ush/python/pygw/src/pygw/timetools.py (hunk 2 above)
Comment (aerorahul) [reply ush/python/pygw/src/pygw/timetools.py:18]: I don't see why not. 
Do you mind if I hold this for another PR where I add some tests for `timetools.py` and then consolidate this?

File: ush/python/pygw/src/pygw/configuration.py (hunk 1 above)
Comment (aerorahul) [reply ush/python/pygw/src/pygw/configuration.py:167]: I understand that, but trying to account and anticipate for "user-error" and correct it is not usually considered safe practice (Don't remember where I read that).  Instead, I would be inclined to add an exception stating that `FAlse` or `trUe` is not an acceptable value.

Also note how these (bools) are cast.
The `if` statement checks if the `string` is in `BOOLS` (Upper and lowercase, as well as the Titlecase).  If the condition is satisfied, the `string` is converted to `lower` to match if it is `True` if the `string` is indeed `True`.

To extend `BOOLS` to account for the typos, would mean to extend `BOOLS` to:
```
BOOLS +=  ['yEs', 'yES', 'yeS', 'YeS']
```
And many more for `True` and `False`, `.TRUE.', `.FALSE'.

File: ush/python/pygw/src/pygw/timetools.py (hunk 2 above)
Comment (WalterKolczynski-NOAA) [reply ush/python/pygw/src/pygw/timetools.py:18]: No objection here.

File: ush/python/pygw/src/pygw/configuration.py (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply ush/python/pygw/src/pygw/configuration.py:167]: Still feels weird that if you call `_true_or_not()` directly on `"TRue"`, you'll get `True`, but if you check whether it is a boolean by looking for it in `BOOLS`, it isn't. Even if the current implementation only calls `_true_or_not()` after first checking `BOOLS`.

Comment (aerorahul) [reply ush/python/pygw/src/pygw/configuration.py:167]: Note I have just moved the methods out of the `class Configuration` to make use of them outside.
Feel free to open a PR to improve/expand the use of typo-ed booleans.

**Comment by aerorahul** [comment]
@CoryMartin-NOAA 
You should be able to use the method from `configuration.py`, `cast_strdict_as_dtypedict` in your PR #1106 as follows:

```
pygfs.task.aero_analysis.AerosolAnalysis(cast_strdict_as_dtypedict(os.environ))
```

It should greatly simply the use and access of the `config` in your scripts.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
When you get a chance, can you add instructions somewhere in the wiki on how to run all the tests manually?

**Comment by aerorahul** [comment]
> When you get a chance, can you add instructions somewhere in the wiki on how to run all the tests manually?

I have updated the `README.md` in the `pygw` to add instructions on how to run the tests manually.
I have tested them on Orion by loading the module `python/3.7.5`.
I don't think I can update the wiki.  Feel free to add these instructions to the appropriate part of the wiki.

**Pull request merged**